### ISSUE TITLE
device number of network interfaces for a vm resource is not assigned in order defined

### ISSUE THREAD

**Issue by njk464** [terra-farm/terraform-provider-xenorchestra#76]
The creation of networks on a vm seems to be ordered by the hashcode of the network string. 

Lines 148-157 in https://github.com/terra-farm/terraform-provider-xenorchestra/blob/master/xoa/resource_xenorchestra_vm.go
```
				Set: func(value interface{}) int {
					network := value.(map[string]interface{})

					macAddress := network["mac_address"].(string)
					networkId := network["network_id"].(string)
					v := fmt.Sprintf("%s-%s", macAddress, networkId)
					log.Printf("[TRACE] Setting network via %s\n", v)

					return hashcode.String(v)
				},
```

This has been causing the device numbers to not be assigned in order of declaration in the terraform file. 

Example

```
resource "xenorchestra_vm" "vyos" {
     memory_max = 4 * var.GB
     cpus = 1

     cloud_config = xenorchestra_cloud_config.bar.template
     name_label = "Vyos"
     name_description = "Vyos"
     template = data.xenorchestra_template.vyos_base.id
     network {
         mac_address = "52:5d:29:d0:c0:f5"
         network_id = "6b2dbe38-f4a7-abde-2be9-32b9a4e2b9c6"
     }
     network {
         mac_address = "52:5d:29:d0:c0:f6"
         network_id = "5655bfa0-1615-9ea5-c0bf-6e2a81549128"
     }
     network {
         mac_address = "52:5d:29:d0:c0:f7"
         network_id = "7b4e8f73-c4a2-87ce-7028-ba884c3aa3bc"
     }
     disk {
         sr_id = "3d96b18e-c71a-e609-8eb7-ed68a379efa0"
         name_label = "Vyos"
         size = 5 * var.GB
     }
 }
```

This vm resource will consistently end up having the following network state. No matter what order they are defined initially

```
"network": [
   {
     "attached": true,
     "device": "0",
     "mac_address": "52:5d:29:d0:c0:f6",
     "network_id": "5655bfa0-1615-9ea5-c0bf-6e2a81549128"
   },
   {
     "attached": true,
     "device": "1",
     "mac_address": "52:5d:29:d0:c0:f5",
     "network_id": "6b2dbe38-f4a7-abde-2be9-32b9a4e2b9c6"
   },
   {
     "attached": true,
     "device": "2",
     "mac_address": "52:5d:29:d0:c0:f7",
     "network_id": "7b4e8f73-c4a2-87ce-7028-ba884c3aa3bc"
   }
 ],
```

Possible solutions that come to mind would be to make the "device" number an optional field in the vm resource's interface definition or to change the Set function above or subsequent code to somehow sort the interfaces by by order defined.

**Comment by ddelnano** [comment]
@njk464 thanks for flagging this issue. When I was working on implementing this I tried a list type (which is ordered) in my first attempt but later switched to a set (which is unordered but can create lists that have deterministic ordering) because it made adding and removing VIFs easier ([source](https://github.com/terra-farm/terraform-provider-xenorchestra/blob/master/xoa/resource_xenorchestra_vm.go#L331-L348)).

I think using a list is the correct thing to do but handling additions and removals will need thought (in addition to making sure user's can seamlessly have their terraform state converted from a set to a list). For example, if we had 3 VIFs (A, B and C) respectively labeled device 1, 2 and 3. If A is deleted should B and C still be device 2 and 3? The XO api doesn't let you change the device number. You can set the device number on creation but not once its created. So if we needed to reorder things that would require deleting and creating new VIFs.

I'm currently working on this but it may take some time for me to figure out what the correct resolution is after evaluating the possible options.

**Comment by ddelnano** [comment]
@njk464 I ended up switching the resource to use a List rather than a set. The details are in #86 and this fix will be available in v0.8.0 of the provider.
